odbc/src/api/functions.rs

```diff
@@ -1279,72 +1270,88 @@ pub unsafe extern "C" fn SQLFetch(statement_handle: HStmt) -> SqlReturn {
-            None => Err(ODBCError::InvalidCursorState),
-        }
-    };
 
-    if let Ok((has_next, warnings_opt)) = move_to_next_result {
-        let mut stmt_attrs = stmt.attributes.write().unwrap();
-        warnings_opt.iter().for_each(|warning| {
-            add_diag_with_function!(
-                MongoHandleRef::from(statement_handle),
-                ODBCError::GeneralWarning(warning.to_string()),
-                function_name.to_string()
-            );
-        });
-        if !has_next {
-            stmt_attrs.row_index_is_valid = false;
-            // No more rows
-            return SqlReturn::NO_DATA;
-        }
-        stmt_attrs.row_index_is_valid = true;
-
-        *stmt.var_data_cache.write().unwrap() = Some(HashMap::new());
-
-        // If there are bound columns, then copy data from the result set into the bound buffers.
-        let mut success_with_info_encountered = false;
-        if let Some(bound_cols) = stmt.bound_cols.read().unwrap().as_ref() {
-            let mongo_handle_for_sql_get_data_helper = MongoHandleRef::from(statement_handle);
-
-            for (col, bound_col_info) in bound_cols.iter() {
-                let sql_return = sql_get_data_helper(
-                    mongo_handle_for_sql_get_data_helper,
-                    *col,
-                    FromPrimitive::from_i16(bound_col_info.target_type).unwrap(),
-                    bound_col_info.target_buffer,
-                    bound_col_info.buffer_length,
-                    bound_col_info.length_or_indicator,
+    let mut success_with_info_encountered = false;
+    let mut global_warnings_opt: Vec<Error> = Vec::new();
+
+    let rowset_size = stmt.attributes.read().unwrap().row_array_size;
+
+    // Use index to figure out which buffer in the array of buffers to use.
+    for index in 0..rowset_size {
+        let move_to_next_result = {
+            let connection = must_be_valid!((*stmt.connection).as_connection());
+            match stmt.mongo_statement.write().unwrap().as_mut() {
+                Some(mongo_stmt) => mongo_stmt
+                    .next(connection.mongo_connection.read().unwrap().as_ref())
+                    .map_err(|e| e.into()),
+                None => Err(ODBCError::InvalidCursorState),
+            }
+        };
+
+        if let Ok((has_next, mut warnings_opt)) = move_to_next_result {
+            let mut stmt_attrs = stmt.attributes.write().unwrap();
+            warnings_opt.iter().for_each(|warning| {
+                add_diag_with_function!(
+                    MongoHandleRef::from(statement_handle),
+                    ODBCError::GeneralWarning(warning.to_string()),
+                    function_name.to_string()
                 );
+            });
+            if !has_next {
+                stmt_attrs.row_index_is_valid = false;
+                // No more rows
+                return SqlReturn::NO_DATA;
+            }
+            stmt_attrs.row_index_is_valid = true;
+
```

Comment:
```suggestion
            if !has_next {
                // No more rows
                return SqlReturn::NO_DATA;
            }

```

I quickly tried SSIS with your branch and things looked good 😄 I had just noticed that we were hanging at this point, and I dont see if/were we use these values elsewhere, so maybe we should just remove the setting of them? (we'll also need to update setting row array size, but otherwise I used the branch as is)

I didn't examine it closely, but it seems to work great. Would be happy to take a closer look if/when this goes into review

Reply:
The problem is probably due to this line `let mut stmt_attrs = stmt.attributes.write().unwrap();` and this line `*stmt.var_data_cache.write().unwrap() = Some(HashMap::new());` after setting `row_index_is_valid`. I think I create 2 write accesses which causes a dead lock.